neqo-common/src/datagram.rs

```diff
@@ -7,20 +7,103 @@
 use std::net::SocketAddr;
 use std::ops::Deref;
 
+use enum_map::Enum;
+
 use crate::hex_with_len;
 
+// ECN (Explicit Congestion Notification) codepoints mapped to the
+// lower 2 bits of the TOS field.
+// https://www.iana.org/assignments/dscp-registry/dscp-registry.xhtml
+#[derive(Copy, Clone, PartialEq, Eq, Enum)]
+#[repr(u8)]
+pub enum IpTosEcn {
+    NotEct = 0b00, // Not-ECT (Not ECN-Capable Transport) [RFC3168]
+    Ect1 = 0b01,   // ECT(1) (ECN-Capable Transport(1))[1] [RFC8311][RFC Errata 5399][RFC9331]
+    Ect0 = 0b10,   // ECT(0) (ECN-Capable Transport(0)) [RFC3168]
+    Ce = 0b11,     // CE (Congestion Experienced) [RFC3168]
+}
+
+impl From<u8> for IpTosEcn {
+    fn from(v: u8) -> Self {
+        match v & 0b11 {
+            0b00 => IpTosEcn::NotEct,
+            0b01 => IpTosEcn::Ect1,
+            0b10 => IpTosEcn::Ect0,
+            0b11 => IpTosEcn::Ce,
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl From<IpTosEcn> for u8 {
+    fn from(val: IpTosEcn) -> Self {
+        val as u8
+    }
+}
+
+impl std::fmt::Debug for IpTosEcn {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            IpTosEcn::NotEct => f.write_str("Not-ECT"),
+            IpTosEcn::Ect1 => f.write_str("ECT(1)"),
+            IpTosEcn::Ect0 => f.write_str("ECT(0)"),
+            IpTosEcn::Ce => f.write_str("CE"),
+        }
+    }
+}
+
+// DiffServ Codepoints, mapped to the upper six bits of the TOS field.
+// https://www.iana.org/assignments/dscp-registry/dscp-registry.xhtml
+#[derive(Copy, Clone, PartialEq, Eq)]
+#[repr(u8)]
+pub enum IpTosDscp {
+    Cs0 = 0b0000_0000,        // [RFC2474]
+    Cs1 = 0b0010_0000,        // [RFC2474]
+    Cs2 = 0b0100_0000,        // [RFC2474]
+    Cs3 = 0b0110_0000,        // [RFC2474]
+    Cs4 = 0b1000_0000,        // [RFC2474]
+    Cs5 = 0b1010_0000,        // [RFC2474]
+    Cs6 = 0b1100_0000,        // [RFC2474]
+    Cs7 = 0b1110_0000,        // [RFC2474]
+    Af11 = 0b0010_1000,       // [RFC2597]
+    Af12 = 0b0011_0000,       // [RFC2597]
+    Af13 = 0b0011_1000,       // [RFC2597]
+    Af21 = 0b0100_1000,       // [RFC2597]
+    Af22 = 0b0101_0000,       // [RFC2597]
+    Af23 = 0b0101_1000,       // [RFC2597]
+    Af31 = 0b0110_1000,       // [RFC2597]
+    Af32 = 0b0111_0000,       // [RFC2597]
+    Af33 = 0b0111_1000,       // [RFC2597]
+    Af41 = 0b1000_1000,       // [RFC2597]
+    Af42 = 0b1001_0000,       // [RFC2597]
+    Af43 = 0b1001_1000,       // [RFC2597]
+    Ef = 0b1011_1000,         // [RFC3246]
+    VoiceAdmit = 0b1011_0000, // [RFC5865]
+    Le = 0b0000_0100,         // [RFC8622]
+}
```

Comment:
This one doesn't have any From implementations.

I wonder whether it would make sense to have an `impl From<u8> for (IpTosDscp, IpTosEcn)`.  You might even have a `type IpTos = (IpTosDscp, IpTosEcn)` alias or a `struct IpTos(u8)` that has appropriate `From` implementations attached.  I guess that's something we'd want if DSCP were more serious, but we need to be realistic.

Reply:
Something like this?
```rust
#[derive(Copy, Clone, PartialEq, Eq, Default)]
pub struct IpTos((IpTosDscp, IpTosEcn));

#[derive(Clone, PartialEq, Eq)]
pub struct Datagram {
    src: SocketAddr,
    dst: SocketAddr,
    tos: IpTos,
    ttl: Option<u8>,
    d: Vec<u8>,
}
```

Would we add a `None` default value to `IpTosDscp` and  `IpTosEcn` then?

Reply:
This works.

The alternative is to have IpTos backed by u8.  Then you would do:

```rust
impl From<IpTosEcn> for IpTos {
  fn from(v: IpTosEcn) -> Self {  Self(u8::from(v)) }
}
impl From<IpTosDscp> for IpTos {
  fn from(v: IpTosDscp) -> Self {  Self(u8::from(v)) }
}
impl From<(IpTosDscp, IpTosEcn)> for IpTos {
  fn from(v: (IpTosDscp, IpTosEcn)) -> Self {  Self(u8::from(v.0) | u8::from(v.1)) }
}
impl From<IpTos> for u8 {
  fn from(v: IpTos) -> Self { v.0 }
}
```

I believe that a `Default` implementation works here with the value equal to zero and that `#[derive(Default)]` does that.

The advantage of that is that you can invoke functions with `IpTos::from(IpTosEcn::ECT1)` and `IpTos::from(IpTosDscp::Cs1)` and `IpTos::from((IpTosDscp::Cs1,IpTosEcn::ECT1))` and `IpTos::from(0x05)` equally.

The disadvantage is that you need a `Debug` implementation that is a little more complex.

```rust
        f.debug_tuple("IpTos")
            .field(&IpTosDscp::from(self.0 & 0xfc))
            .field(&IpTosEcn::from(self.0 & 0x3))
            .finish()
```